File: src/boj/boj_1916.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class boj_1916 {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        LinkedList<Edge>[] graph = new LinkedList[n];
        for (int i = 0; i < n; i++) {
            graph[i] = new LinkedList<>();
        }

        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int from = Integer.parseInt(st.nextToken()) - 1;
            int to = Integer.parseInt(st.nextToken()) - 1;
            int weight = Integer.parseInt(st.nextToken());
            graph[from].add(new Edge(to, weight));
        }
        StringTokenizer st = new StringTokenizer(br.readLine());
        int s = Integer.parseInt(st.nextToken()) - 1;
        int e = Integer.parseInt(st.nextToken()) - 1;

        System.out.println(dijkstra(n, graph, s, e));
        br.close();
    }

    static int dijkstra(int n, LinkedList<Edge>[] graph, int s, int e) {
        int[] distance = new int[n];
        Arrays.fill(distance, Integer.MAX_VALUE);
        distance[s] = 0;
        PriorityQueue<Integer> pq = new PriorityQueue<>(Comparator.comparingInt(i -> distance[i]));
        pq.add(s);
        boolean[] visited = new boolean[n];

        while (!pq.isEmpty()) {
            int current = pq.poll();
            if (current == e) {
                return distance[e];
            }
            if (visited[current]) {
                continue;
            }
            visited[current] = true;
            for (Edge edge : graph[current]) {
                if (distance[current] + edge.weight < distance[edge.to]) {
                    distance[edge.to] = distance[current] + edge.weight;
                    pq.add(edge.to);
                }
            }
        }
        return -1;
    }

    static class Edge {
        int to;
        int weight;

        public Edge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }
}

File: src/boj/boj_1916.kt
package boj_1916

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

var n = 0
var m = 0
lateinit var buses: Array<LinkedList<Pair<Int, Int>>>
lateinit var distance: IntArray
var start = 0
var destination = 0

fun main() {
    init()
    dijkstra()
    println(distance[destination])
}

fun init() = with(BufferedReader(InputStreamReader(System.`in`))) {
    n = readLine().toInt()
    buses = Array(n) { LinkedList<Pair<Int, Int>>() }
    distance = IntArray(n) { Int.MAX_VALUE }
    m = readLine().toInt()
    repeat(m) {
        with(StringTokenizer(readLine())) {
            val u = nextToken().toInt() - 1
            val v = nextToken().toInt() - 1
            val w = nextToken().toInt()
            buses[u].add(Pair(v, w))
        }
    }
    with(StringTokenizer(readLine())) {
        start = nextToken().toInt() - 1
        destination = nextToken().toInt() - 1
    }
    close()
}

fun dijkstra() {
    val priorityQueue = PriorityQueue<Pair<Int, Int>>() { n1, n2 ->
        n1.second.compareTo(n2.second)
    }.apply { add(Pair(start, 0)) }
    distance[start] = 0

    while (priorityQueue.isNotEmpty()) {
        val current = priorityQueue.poll()
        if (current.first == destination) {
            return
        }

        for (next in buses[current.first]) {
            if (current.second + next.second < distance[next.first]) {
                distance[next.first] = current.second + next.second
                priorityQueue.add(Pair(next.first, distance[next.first]))
            }
        }
    }
}

